### DISCUSSION TITLE
How to: Manage Actions Consumption and Cost

### DISCUSSION THREAD

**Discussion by KittyChiu** [community/community#129964] [Discover: GitHub Best Practices]
> [!NOTE]
> TL;DR Got a bill shock? Here's a value stream approach to analyse Actions minutes consumption. First, know your CICD value stream. If none, it's time to define it. Next, identify where along the flow are Actions involved and gather the data. Finally, do the math to forecast and optimise your next monthly consumption.

GitHub has per seat license where the cost is pretty predictable with headcounts. However as we adopt DevSecOps practices with GitHub Actions, one of the top challenges most organisations will encounter is the metered billing. From CICD to Security features, they all leverage Actions. Some GitHub platform owner may either get a 'bill shock' for this line item, or seeing their included Actions minutes vaporise rapidly within the first few days of the billing month, or being conservative and run CICD in their own servers (or under their desk).

These are signs of not managing Actions consumption early and pragmatically. Fear, uncertainty and doubt (FUD) may cause undesired behaviours, and may manifest as compromising validation checks that may cripple the product quality and reliability.

## ⚙️  Apply FinOps with Value Stream

Applying [FinOps](https://www.finops.org/framework/) should be considered when adopting metered services like Actions. In a nutshell, the FinOps is about

- Moving away from siloed roles and responsibilities, to shared accountability and responsibility
- Collaborative effort to optimise between business and IT
- Make decisions base on business value rather than COGS or ROI

Three adoption phases of FinOps all start with **visibility to usage**. Understanding Actions usage is the fundamental step.

Having a structure approach to gain usage visibility will demystify the ambiguity, enable proper accountability, and allow better budgeting and forecasting to practice DevSecOps. Here are the four key steps:

## 🏗️  1. Define your CI/CD value stream and practices

The first step is to know your value stream. Quality should be well-planned and intentional.

_For example_, trunk-based development may typically have a CI/CD value stream that looks like this:

```mermaid
%%{init: { 'logLevel': 'error', 'theme': 'forest' , 'themeVariables': {
              'cScale0': '#CCC',
              'cScale1': '#CCC',
              'cScale2': '#CCC',
              'cScale3': '#CCC',
              'cScale4': '#CCC',
              'cScale5': '#CCC'
       } } }%%
timeline
    A feature prioritised: auto-backlog management : agile planning : create feature branch
    Development: commits against branch : unit test: lint : static application security testing : software composition analysis
    Continuous integration : create pull request : build : integration test : static application security testing : software composition analysis : peer review : merge into default branch
    Continuous delivery : build : architecture fitness : package : sign & publish : deploy to non-prod : manual & exploratory test : dynamic application security testing
    Deploy to production : deployment gates : canary deployment : blue / green deployment
    Continuous monitoring : detect : response : post-mortem : live-site review

```

## 🧮  2. Gather Actions usage data

Next step is to trace Actions usage for the identified activities. The activities may be related to first-party GitHub Actions, third-party from the [Marketplace](https://github.com/marketplace?type=actions) or other open-sourced community. The data we want to trace are:

- `average duration` - from past success workflow runs
- `success rate` - complete and accurate rate
- `frequency` - recorded occurrence

_For example_,

| Stage                        | Trigger                         | Actions <br />activities                                                                                                                                                                                 | Duration to <br />complete stage | Success <br />rate |   Frequency    |
| ---------------------------- | ------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------------------------------: | :----------------: | :------------: |
| A feature <br />prioritised  | Scheduled                       | auto-backlog management[^issue][^project]                                                                                                                                                                |                1m                |        80%         |     weekly     |
| Development                  | Commits to <br />feature branch | commit, <br />build, <br />unit test, <br />lint, <br /> dependency scanning[^dependency-review], <br />secret scanning[^push-protection]                                                                |                2m                |        50%         | 15 times / day |
| Continuous <br />integration | Merge to <br />default branch   | build, <br />integration test, <br />static application security testing scanning[^codeql], <br />dependency review[^dependency-review]                                                                  |               20m                |        70%         | 5 times / day  |
| Continuous <br />delivery    | Generate <br />a release        | build, <br />architecture fitness test, <br />package, <br />sign[^sign] & publish[^release][^packages], <br />deploy to environments, <br />exploratory test,<br />dynamic application security testing |               20m                |        80%         |     weekly     |
| Deploy to <br />production   | Request to <br />go-live        | deployment gates[^deployment-gates], <br />deploy to production                                                                                                                                          |               20m                |        90%         |     ad-hoc     |
| Continuous monitoring        | Scheduled                       | dependency scanning[^dependabot]                                                                                                                                                                         |                1m                |        100%        |     daily      |

## 💻  3. Evaluate Actions minutes usage

Third step is to evaluate the Actions durations with the data gathered.

**Calculation**,

> $$ \text{ Total activity duration } = \text{ Duration } \div \text{ Success Rate } \times \text{ Frequency } $$

Evaluate the rest of the stages, and the visual may show more clearly where and how are Actions used in your CICD:

```mermaid
%%{init: { 'logLevel': 'error', 'theme': 'forest' , 'themeVariables': {
              'cScale0': '#888',
              'cScale1': '#CCC',
              'cScale2': '#CCC',
              'cScale3': '#CCC',
              'cScale4': '#CCC',
              'cScale5': '#CCC',
              'cScale6': '#CCC'
       } } }%%
timeline
    Stage: Average run duration : Success rate : Frequency : Normalised usage per week
		A feature prioritised: 1m : 80% : weekly : 1m
    Development: 2m : 50% : 15 times per day : 300m
    Continuous integration : 20m : 70% : 5 times per day : 714m
    Continuous delivery : 20m : 80% : weekly : 25m
    Deploy to production : 20m : 90% : ad-hoc : 22m
    Continuous monitoring : 1m : 100% : daily : 7m
```

> [!TIP]
> I have open-sourced a GitHub action to help implementing what's been discussed here. There are example use cases of how to apply in workflows (yml) - [do check it out](https://github.com/KittyChiu/workflow-metrics)!

## 📐  4. Adjustment for forecasting - growth rate and confidence level

Finally, the calculation can be extended to forecast future consumption. To do so, factor in these variables to adjust:

- **Growth rate** is expected growth including increase in number of developers, teams and/or repositories
- **Confidence level** is the success rate used to estimate an interval. Then the confidence interval is an estimated range based on historical data, incorporating the risk of uncertainty and unplanned factors

Forecasting may not 100% accurate at first calculation. It is not _set once and forget_, but rather it requires continuous refinement.

## 🔄  'Inspect and adapt' is the flywheel

Visibility to Actions usage in value stream is the first step to gain assurance in Actions consumption. It needs to be followed by iterative process of _inspect and adapt_ by both business and engineering. For FinOps framework, this is the [Inform, Optimize, and Operate activities](https://www.finops.org/framework/maturity-model/) which may look like:

- _Continuously monitor_ with live-dashboards accessible to all responsible individuals (including engineers). Using budget alerts for 50%, 75% and 90% threshold, and _delegate_ responsibility to respond to unplanned spikes.
- Weekly _inspect_ the level of Actions consumption by the product owner and engineering, and _showback_ accountability to the business/service owners.
- Monthly service view to _inspect_ where the Actions minutes are being used help to optimize your CICD, and _remediate_ technical debts. Pay attention to data such as workflow runs that continue to fail or redundant.

Our ultimate goal is to minimise surprises and manage FUD for all stakeholders, make practicing DevSecOps sustainable and value adding to business.

[^issue]: [Managing Issues with Actions](https://docs.github.com/en/actions/managing-issues-and-pull-requests/adding-labels-to-issues)
[^project]: [Using Actions for project management](https://docs.github.com/en/actions/managing-issues-and-pull-requests/using-github-actions-for-project-management)
[^push-protection]: [Protecting pushes with secret scanning](https://docs.github.com/en/code-security/secret-scanning/protecting-pushes-with-secret-scanning)
[^dependency-review]: [Dependency review](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-dependency-review)
[^codeql]: [Code scanning with CodeQL](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/about-code-scanning)
[^sign]: [Artifact attestations](https://docs.github.com/en/actions/security-guides/using-artifact-attestations-to-establish-provenance-for-builds)
[^release]: [Managing releases in a repository](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository)
[^packages]: [Publish and install with Actions](https://docs.github.com/en/packages/managing-github-packages-using-github-actions-workflows/publishing-and-installing-a-package-with-github-actions)
[^deployment-gates]: [Custom deployment protection rules](https://docs.github.com/en/actions/deployment/protecting-deployments/creating-custom-deployment-protection-rules)
[^dependabot]: [Use Dependabot with Actions](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/automating-dependabot-with-github-actions)

**Comment by joaodadas** [comment]
This help me a lot in my study

**Comment by 09170988219** [comment]
من میخوام مجموعه دارایی رایکجابیاورم ازان استفاده کنم
